Airflow Adapter

DomMarq opened the pull request:
The Airflow Adapter, with basic functionalities, plus:
- [x] Release names are different from release IDs (waiting on [Emilia's PR#34](https://github.com/istio-releases/release-ui/pull/34))
- [x] Ability to connect to the production Airflow SQL database (waiting on perms)
- [x] Links to different repos on the details page

This pushes filtering and sorting to the Adapter portion, and allows for connection to the data dumped SQL database in the istio-release-ui GCP project at the moment.

Commit:
Began work on airflow adapter - added SQL query constructor for releases
Allows for filtering and sorting of releases by the SQL database rather
than the release-ui server

Commit:
Moving airflow-db constants to init

Commit:
Added get release and get task query generators

Commit:
changed how the connector will work

Commit:
Made sure requesting works

Commit:
Merge remote-tracking branch 'origin' into airflow-adapter

Commit:
Scheduled release info showing up on front page
Manual not being provided all info for some reason

Commit:
Merge branch 'master' into airflow-adapter

Commit:
Merge branch 'master' into airflow-adapter

Commit:
Merge branch 'airflow-adapter' of https://github.com/DomMarq/release-ui into airflow-adapter

Commit:
Rename "type" (builtin function)

Commit:
incorporate release_type, branch. Rename "type" (builtin function)

Commit:
Kind of fixed no tasks bug(more of a band-aid)

Commit:
Merge branch 'airflow-adapter' of https://github.com/DomMarq/release-ui into airflow-adapter

Commit:
Kind of fixed no tasks bug(more of a band-aid)

Commit:
Merge branch 'airflow-adapter' of https://github.com/DomMarq/release-ui into airflow-adapter

Commit:
Merge remote-tracking branch 'origin/master' into airflow-adapter

Commit:
Created a way to uniquely id a dag run

Commit:
Got tasks showing up on details page about ~50% of the time

Commit:
Tasks show up about 80% now. Still not sure why this is an issue

Commit:
Tasks show up about 80% now. Still not sure why this is an issue

Commit:
Merge branch 'airflow-adapter' of https://github.com/DomMarq/release-ui into airflow-adapter

Commit:
fake data now the same as release-ui

Commit:
added release.release_id to from_sql, got App Engine deployment working

Commit:
Tidy up, add comments

DomMarq commented:
Note that while this is not quite ready for merging, I wanted to put it out there for review sooner rather than later, especially considering what a big PR it is.

sebastienvas commented on airflow_adapter.py:
```diff
@@ -0,0 +1,119 @@
+  def __init__(self, airflow_db):
+    self._airflow_db = airflow_db
+
+    self._releases = self._airflow_db.query('SELECT * FROM dag_run;')
```
why do we need this ?

DomMarq replied:
It is there so that we can assemble branch and type information at the start, rather than doing it for each and every request. I suppose we could move these to get_branches() and get_types, however.

sebastienvas replied:
I understand this, but those values will change as we create more releases, and therefore instead of doing it in the constructor you should do it in a different methods that you can call when your cache is invalidated.

sebastienvas commented on airflow_adapter.py:
```diff
@@ -0,0 +1,119 @@
+      types.add(self._releases[release].release_type)
+    self._types = list(types)
+
+  def get_releases(self, start_date, end_date, datetype, state,
```
I think passing a filter object would be easier to use here since we are updating the interface. If you need to initialize this, you should create a method for this and cache it. I am assuming that calling get_labels, get_branches, etc... should use the cache, but that cache should probably be invalidated once in a while. It could be a background job, or to start just you would just invalidate the cache when any of get_branches is called.

DomMarq replied:
So just to check, the filter object should function in a similar manner to the release object, and then the filter conditions are called like this:
```Python
start_date = filter_conditions.start_date
datetype = filter_conditions.datetype
```
?

sebastienvas replied:
Yes. I would call it filter_options.

sebastienvas commented on airflow_connector.py:
```diff
@@ -0,0 +1,101 @@
+      cursor.execute(request)
+      response = cursor.fetchall()
+    except MySQLdb.OperationalError, e:
+      print '*****************ERROR TRIGGERED WITH MYSQLDB*********'
```
use logging.

sebastienvas commented on airflow_connector.py:
```diff
@@ -0,0 +1,101 @@
+    # even if the db connection has been lost
+    print request
+    try:
+      cursor = self._airflow_db.cursor()
```
do you expect self._airflow_db.cursor() to raise an exception ? if not only put the methods raising an exception in the try catch

DomMarq replied:
I originally had it outside, but it has raised exceptions before when the connection expires, so I moved it in.

sebastienvas commented on airflow_connector.py:
```diff
@@ -0,0 +1,101 @@
+      print 'Error: '
+      print e
+      print ''
+      if e[0] == 2006:
```
duplicating code here,
```python
if e[0] in [2006, 2013]:
```

sebastienvas replied:
And also use the Error code from the mysql libary ie MYsqldb.<NAME of error code>, instead of just using a number (that has no meaning to me :))

DomMarq replied:
So MySQL gives me this when it raises an error, as an example:
(2006, 'MySQL server has gone away')
I'm not sure what you mean by the Error code. I found [this](https://dev.mysql.com/doc/refman/8.0/en/error-messages-server.html) and [this](https://dev.mysql.com/doc/refman/8.0/en/error-messages-client.html), which seems to say that what you are asking for doesn't exist?

sebastienvas commented on airflow_connector.py:
```diff
@@ -0,0 +1,101 @@
+      print ''
+      if e[0] == 2006:
+        self._airflow_db = self.create_connection()
+        cursor = self._airflow_db.cursor()
```
why creating a cursor to close it after ?

Commit:
Merge remote-tracking branch 'origin' into airflow-adapter

Commit:
Requested Changes:
- added logging
- filter object
- some smaller changes

Commit:
Merge remote-tracking branch 'origin' into airflow-adapter

sebastienvas commented on adapters/adapter_abc.py:
```diff
@@ -10,7 +10,7 @@ class Adapter(object):
 
   @abc.abstractmethod
   def get_releases(self):
```
you need to change the arguments here as well and document them as well as the return values.

sebastienvas replied:
Please update this with:
```python
def get_releases(self, filter_options):
```
and Add a Args section to the comment.

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,125 @@
+    """
+    return self._types
+
+  def update_cache(self):
```
please make this private.

sebastienvas replied:
please use a lock here.

DomMarq replied:
I'm a little confused about what a lock is, and how I would use one?

sebastienvas replied:
In your constructor:
```python
def __init__(self):
  self._lock = threading.Lock()
```
in your code each time you update or read data:
```python
with self._lock:
   self._releases ....

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,125 @@
+    types = set()
+    for release in self._releases:
+      types.add(self._releases[release].release_type)
+    self._types = list(types)
```
Remove this and call self._update_cache()

sebastienvas commented on adapters/airflow_connector.py:
```diff
@@ -1,11 +1,12 @@
   """"Provides the methods which allow interaction with the Airflow SQL database."""  # pylint: disable=line-too-long
 
-  def __init__(self, unix_socket, host, user, password, db):
+  def __init__(self, user, password, db, host=None, unix_socket=None):
```
please document parameters

sebastienvas commented on adapters/airflow_connector.py:
```diff
@@ -24,33 +25,60 @@ def query(self, request):
-    cursor = self._airflow_db.cursor()
     # The following ensures that the query executes and returns,
     # even if the db connection has been lost
+    logging.info(request)
```
When logging, you might want to give some more information. Just dumping the data is usually not really useful. 
```python
logging.info("Running query %s against Airflow DB", request)
```

sebastienvas commented on adapters/airflow_connector.py:
```diff
@@ -24,33 +25,60 @@ def query(self, request):
     except MySQLdb.OperationalError, e:
-      if e[0] == 2006:
+      logging.error('Error %i triggered with MySQLdb: %s'  % (e[0], e[1]))
+      if e[0] in [2006, 2013]:
```
Errors must be declared globally on the mysql module, please use the global variable instead of those numbers. 
```python
from mysql.connector import errorcode
...
if e[0] in [errorcode.<>, errorcode.<>]
```

sebastienvas commented on adapters/airflow_connector.py:
```diff
@@ -24,33 +25,60 @@ def query(self, request):
 
     return response
 
   def create_connection(self):
```
please make this private.

sebastienvas commented on adapters/airflow_connector.py:
```diff
@@ -24,33 +25,60 @@ def query(self, request):
+          db=self._db)
     return db
 
+  def check_connection(self):
```
please make this private

sebastienvas replied:
this method does not seem to be used

sebastienvas commented on adapters/to_sql.py:
```diff
@@ -0,0 +1,122 @@
+    state: string format which follows the same format in the SQl db (str)
+  """
+  state = int(state)
+  if state == 0:
```
do not use number, but use the constant representing the number to make it easier to read.

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,187 @@
+  Returns:
+    int
+  """
+  if state == 'none':
```
same comment here, use the constant representing the number. You might want to define a global map for this.
```python
STATE_FROM_STRING = {
"running": State.RUNNING,
...
}
```
and then use it like:

STATE_FROM_STRING.get(state, STATE.UNKNOWN)

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,187 @@
+from resources.to_timestamp import to_timestamp
+
+
+def read_releases(release_data, airflow_db):
```
why don't you move all the methods that takes in an airflow db to the AirflowDB class ?

sebastienvas replied:
can we do a join select instead ? would it be possible to get all the information in one query ?

DomMarq replied:
Sure, I can move them into the AirflowDB class, though I feel as if it might make more sense to have those that just take in airflow_db just so that it can pass it off to other functions in the AirflowAdapter class?

DomMarq replied:
I'm not too well-versed in SQL, but it does look like it might be possible to get all of the info in one query. It would also push more of the data retrieval onto the SQL server, so I'll go ahead and implement it.

sebastienvas replied:
Maybe that s a little bit too soon to do that, but eventually what I want to get is getting the raw data from Airflow is one thing, and converting the data is another. Let's do that in a future PR once we have all the required data.

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,187 @@
+    started = to_timestamp(item[2])
+    task_ids, most_recent_task, state = get_task_info(started, airflow_db)
+    xcom_dict, green_sha = get_xcom(started, item[1], airflow_db)
+    release.release_id = item[1] + '@' + str(item[2])
```
please use http://mysql-python.sourceforge.net/MySQLdb-1.2.2/public/MySQLdb.cursors.DictCursor-class.html instead if using indices.

Commit:
Requested Changes #2, added ability to connect to real data on AppEngine

Commit:
deleted state_from_string() - not needed anymore

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,125 @@
+  def _update_cache(self):
+    with self._lock:
+      raw_release_data = self._airflow_db.query('SELECT dag_id, execution_date FROM dag_run;')
+      self._releases = read_releases(raw_release_data, self._airflow_db)
```
it would be good to move raw_release_data = self._airflow_db.query('SELECT dag_id, execution_date FROM dag_run;') out of the lock, and to only lock when updating the resource as this could create some contention and performance issue.

sebastienvas replied:
```python
  raw_release_data = self._airflow_db.query('SELECT dag_id, execution_date FROM dag_run;')
  releases = read_releases(raw_release_data, self._airflow_db)
  branches = set()
  types = set()
  for release in releases:
    branches.add(releases[release].branch)
    types.add(releases[release].release_type)

  with self._lock:
    self._releases = releases
    self._branches = list(branches)
    self._release_types = list(types)
    self._cache_last_updated = datetime.now()
```

sebastienvas commented on adapters/airflow_connector.py:
```diff
@@ -24,34 +35,45 @@ def query(self, request):
     return response
 
-  def create_connection(self):
+  def _create_connection(self):
```
Do you have to return the db ? or can you just update self._airflow_db directly ?

DomMarq replied:
Now updates variable directly

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,125 @@
+      Array of branches as strings.
+    """
+    # check if cache is older than 30 mins. If true, then update cache
+    if (self._cache_last_updated - datetime.now()).total_seconds() > 1800:
```
Please set 1800 in global variable. like CACHE_TTL = 1800

Commit:
Fixed bug which caused incorrect tasks to render in details page

Commit:
Requested Changes #3

Commit:
Small comment change

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,127 @@
+
+  def _update_cache(self):
+    raw_release_data = self._airflow_db.query('SELECT dag_id, execution_date FROM dag_run;')
+    self._releases = read_releases(raw_release_data, self._airflow_db)
```
it seems self._releases is not used, so please rename it to be a local variable for this method. ie releases

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,127 @@
+    Returns:
+      Array of types as strings.
+    """
+    if (self._cache_last_updated - datetime.now()).total_seconds() > CACHE_TTL:
```
please move the if into self._update_cache() and make the method return if the cache is still valid.

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+    else:
+      task.last_modified = to_timestamp(item.end_date)
+    task.log_url = 'https://youtu.be/dQw4w9WgXcQ'  # TODO(dommarques): figure out how to get the log in here
+    if item.state == None:
```
move this before 
if item.state == None:
￼ +      task.status = STATE_FROM_STRING.get('none')

that way you don't need an if

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+    if item.state == None:
+      item.state == 'none'
+    else:
+      task.error = item.state
```
do you want task.error to be set if state is successful  or running ?

DomMarq replied:
It's totally up to you. I originally had it there just to make sure that the task error was always reading, but since we know it works now it doesn't need to be that way anymore

sebastienvas replied:
I think task error should be a summary of the error. Do we have this data in Airflow ? @rkpagadala

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+  state = 0
+  task_ids = []
+  if task_objects:
+    most_recent_task = task_objects[len(task_objects) - 1]
```
inpython you can just do task_objects[- 1]

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+  # occasionally, the xcom data doesn't exist yet. This is a workaround
+  xcom_named_tuple = namedtuple('Row', ['xcom_dict', 'green_sha'])
+  try:
+    xcom_data = xcom_named_tuple._make(xcom_data)
```
can you just do
xcom_data = xcom_named_tuple(*xcom_data) ?

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+  try:
+    xcom_data = xcom_named_tuple._make(xcom_data)
+    # The index 0 is here because the data returns in a tuple
+    xcom_dict = json.loads(xcom_data.xcom_dict[0])
```
What does 0 represent ? Could you please define a global variable key ?

DomMarq replied:
The 0 is there only because the data returns in a tuple with only one entry, for example:
> xcom_dict = (some_data,)

Because of this, I didn't think a global variable key was necessary. I can make one if you would like

sebastienvas replied:
what is the other part of the tuple ? can we have get_xcom() method get rid of unused data ?

DomMarq replied:
There is no other part of the tuple, it is essentially a single variable wrapped in a tuple. I have this in read_xcom_vars() because the main function of this method is to get the data from the SQL query into a more usable format, which is essentially what this is doing in my mind.

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+    xcom_data = xcom_named_tuple._make(xcom_data)
+    # The index 0 is here because the data returns in a tuple
+    xcom_dict = json.loads(xcom_data.xcom_dict[0])
+    green_sha = xcom_data.green_sha[0]
```
Same comment here.

DomMarq replied:
See above

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+    release.links = construct_links(green_sha)  # TODO(dommarques) these need to be implemented into airflow first, or we make our own way to get the links pylint: disable=line-too-long
+    release.labels = [item.dag_id]
+    release.state = state
+    release.branch, release.release_type = parse_dag_id(item.dag_id)
```
I thought we agreed to use the xcom for that ?

DomMarq replied:
We did, I forgot to change that. Fixing now

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+      # allows for continuation, even if xcom has not been generated yet
+      release.name = release.release_id
+    else:
+      release.name = xcom_dict['VERSION']
```
I guess you could overwrite release_branch and release_type from xcom data

sebastienvas replied:
Use a global Variable for 'VERSION'. If there are more keys from XCOM, please use a Class. If the key changes in the future, we can just update the class, and not looking for all instance of a key in the file.

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+  for i, char in enumerate(dag_id):
+    if char == '_':
+      underscore_count += 1
+    if underscore_count == 1 and not last_underscore:
```
it seems this statement does not do anything. Unless you have an indentation issue at line 186. Note that this method does not return anything if underscore_count != 2

sebastienvas replied:
Have you looked at str.find and str.partition or str.split, this might help you

sebastienvas replied:
I think you can just do 
```python
info = dag_id.split('_')
if len(info) in [2, 3]:
  return info[-2], info[-1]
# handle not valid type
```

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+    release.labels = [item.dag_id]
+    release.state = state
+    release.branch, release.release_type = parse_dag_id(item.dag_id)
+    if xcom_dict is None:
```
is this possible ? don't you need a XCOM to start a release @rkpagadala ?

DomMarq replied:
I didn't think it was possible either, but I kept running into this problem with certain releases, not really sure why but this was the only fix. They were the same releases where the tasks didn't exist for the release yet.

sebastienvas replied:
Can we ignore those release in that case ? ie do not save or return them.

DomMarq replied:
Sure, that should be fairly simple

sebastienvas commented on adapters/to_sql.py:
```diff
@@ -0,0 +1,120 @@
+  end_date = datetime.datetime.fromtimestamp(filter_options.end_date)
+  # append the date filter to the query based on:
+  # datetype, start_date, and end_date
+  if filter_options.datetype == 0:
```
it is the same outcome if datetype == 0 or not. Is this expected ?

sebastienvas replied:
it might be worthwhile to look at jinja for query templating as this might be easier to read. Maybe in a later PR.

DomMarq replied:
I think my original idea was to be able to filter by last modified, but since this is dependent on tasks this turns out to have no significance anymore. Deleting

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,194 @@
+                     'failed': State.FAILED,
+                     'shutdown': State.ABANDONED,
+                     'upstream_failed': State.PENDING}
+
```
```python
STRING_FROM_STATE =  {v: k for k, v in STATE_FROM_STRING.iteritems()}
```

sebastienvas commented on adapters/to_sql.py:
```diff
@@ -0,0 +1,120 @@
+  Returns:
+    state: string format which follows the same format in the SQl db (str)
+  """
+  state = int(state)
```
Use STRING_FROM_STATE

sebastienvas commented on data/filter_options.py:
```diff
@@ -0,0 +1,116 @@
+
+  @state.setter
+  def state(self, value):
+    if self._validate_type(int, value, 'state'):
```
I don't think this is required. Maybe add a is_valid() to the State class:
```python
def is_valid(self, state):
  return 0 <= state <= 4

sebastienvas commented on resources/to_timestamp.py:
```diff
@@ -0,0 +1,20 @@
+    return int(current_timestamp())
+
+
+def current_timestamp():
```
return int(time.time())

sebastienvas replied:
you usually don t need a method for this.

sebastienvas commented on resources/to_timestamp.py:
```diff
@@ -0,0 +1,20 @@
+
+
+def to_timestamp(datetime_object):
+  logging.debug('datetime: ' + str(datetime_object))
```
same comment here you ususally don t need a method for that:
time.mktime(dt.timetuple())

sebastienvas reviewed (changes requested):
Please take a look

Commit:
Requested Changes, Fixed lack of reliability on details page

Commit:
Delete to_timestamp.py

Commit:
Forgot xcom_keys class

Commit:
Added real links on details page

Commit:
- Fixed bug which kept the most recent task from showing up correctly
- Fixed bug which caused a release to show up as successful if pending
and not failed(hopefully)
- Fixed last_recent_task not being correct when release fails

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,133 @@
+    with self._lock:
+      return self._release_types
+
+  def _update_cache(self):
```
Prefer the other way:
if valid:
  return

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,133 @@
+
+    return release_data
+
+  def get_tasks(self, dag_id, execution_date):
```
I am thinking that execution_data is really airflow implementation and we should not bring that to the interface, can we use release_id instead, and do a mapping inside the AirflowAdapter impl ?

DomMarq replied:
Sure I'll make the change

sebastienvas commented on adapters/from_sql.py:
```diff
@@ -0,0 +1,214 @@
+def construct_links(green_sha):
+  """Takes the green build sha and derives the repo links from that."""
+  response = []
+  # green_sha[1:-1] gets rid of the quotes that enclose the SHA
```
use 
```python
green_sha = green_sha.replace("'", "").strip() 
```

sebastienvas commented on data/filter_options.py:
```diff
@@ -0,0 +1,119 @@
+  @state.setter
+  def state(self, value):
+    if self._is_valid_state(value):
+      if value == 0:
```
you don't need to compare every value you can just set it:
```python
if not self._is_valid_state:
  raise ValueError('Invalid input for status')
self._filter_options['state'] = state
```

sebastienvas commented on main.py:
```diff
@@ -11,28 +11,33 @@
+else:
+  # Connect using the unix socket located at
+  # /cloudsql/cloudsql-connection-name.
+  cloudsql_connection_name = os.environ.get('CLOUDSQL_CONNECTION_NAME')
```
unless you define those in a method, please use ALL_CAPS for global variable.

sebastienvas commented on main.py:
```diff
@@ -11,28 +11,33 @@
+# creates the proper database connection depending on whether the app is
+# deployed locally or on App Engine
+if os.getenv('SERVER_SOFTWARE', '').startswith('Development/'):
+  airflow_db = AirflowDB(host='35.193.234.53',
```
The Host should also be passed as an env.

sebastienvas reviewed (changes requested):
MInor changes, please let me know when you are done with those.

Commit:
Requested Changes

Commit:
Made the most recent task be the failed task on failed releases

Commit:
Dag name parsing now handled in the airflow adapter

Commit:
Fixed a bug - "Execution date referenced before assignment"

Commit:
Merge branch 'master' of https://github.com/istio-releases/release-ui into airflow-adapter

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,141 @@
+
+    return releases_data
+
+  def get_release(self, release_name):
```
PLease update the documentation

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,141 @@
+    return release_data
+
+  def get_tasks(self, dag_name):
+    """Retrieve all task information.
```
can we use something else than dag_name, would release_id make more sense ? Also please update the doc to match the name of variable.

DomMarq replied:
sure i'll make that change

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,141 @@
+
+    return task_objects
+
+  def get_task(self, task_name, dag_name):
```
Same comment here.

Commit:
Requested changes
- renamed release_name to release_id
- updated pydocs in AirflowAdapter

sebastienvas commented on adapters/airflow_adapter.py:
```diff
@@ -0,0 +1,158 @@
+  def _update_cache(self):
+    if (datetime.now() - self._cache_last_updated).total_seconds() < CACHE_TTL:
+      return
+    else:
```
you don't need else if you return

Commit:
delete else statement in _update_cache()
